### DISCUSSION TITLE
How can i get refresh token after a user goes through google oauth?

### DISCUSSION THREAD

**Discussion by starascendin** [pocketbase/pocketbase#6104] [Q&A]
So right now i've integrated w/ google oauth like:
```
			const authData = await pb.collection('users').authWithOAuth2({
				provider: 'google',
				scopes: [
					'https://www.googleapis.com/auth/userinfo.profile',
					'https://www.googleapis.com/auth/userinfo.email',
					'https://www.googleapis.com/auth/youtube',
					'https://www.googleapis.com/auth/youtube.readonly'
				],
				access_type: 'offline',
				prompt: 'consent',
				redirectUrl: 'http://localhost:8090/api/oauth2-redirect'
			});
```

but in the meta of authData, i only see accessToken and not refreshToken. What's the way to retrieve refresh token?

**Comment by ganigeorgiev** [comment]
`access_type` and `prompt` query params have no effect on the constructed provider auth url and if you want to add custom query params to it you'll have to define a cusom `urlCallback` function and modify it from there (this also allows more control over the default popup/tab opener). For example:

```js
const authData = await pb.collection('users').authWithOAuth2({
    provider: 'google',
    scopes: [
        'https://www.googleapis.com/auth/userinfo.profile',
        'https://www.googleapis.com/auth/userinfo.email',
        'https://www.googleapis.com/auth/youtube',
        'https://www.googleapis.com/auth/youtube.readonly'
    ],
    urlCallback: (rawUrl) => {
        let url = new URL(rawUrl);
        url.searchParams.set('access_type', 'offline');
        url.searchParams.set('prompt', 'consent');
        // ... any other query param ...

        // Open in new tab or as popup based on the desired behavior...
        //
        // As a side-note since you are defining a custom urlCallback,
        // be aware that Safari may block window.open when executed in async handler
        // and you may have to open the window before the oauth2 call
        // (see https://github.com/pocketbase/pocketbase/discussions/2429#discussioncomment-5943061).
        window.open(url.toString(), "_blank", "");
    }
});
```

**Reply by starascendin** [reply]
awesome, thanks for this clarification!

**Reply by chnfyi** [reply]
Note that `prompt` is a v2 and above endpoint parameter. PocketBase currently use v1 endpoint so for your information `prompt=consent` v1 equivalent is `approval_prompt=force`.

**Reply by ganigeorgiev** [reply]
@chnfyi Could you provide a link to the related Google docs? The v1 API I think it is currently only for the userinfo endpoint and the above params are set for the auth url.

**Reply by ganigeorgiev** [reply]
The only documentation that I could find related to the auth endpoint versioning was this https://stackoverflow.com/questions/33137140/googles-oauth2-endpoint-v1-vs-v2-changes/33156346 answer by a Googler, which indeed mentions that there is a newer version for the `/auth` endpoint for which I assumed that if you don't specify versioning you'll use the latest one and based on that answer this doesn't seem to be the case (_note also that the same non-versioned auth url is what the default Go [oauth2/google.Endpoint.AuthURL](https://github.com/golang/oauth2/blob/master/google/google.go#L26) is using_).

It is very confusing and I'm having difficulties navigating/searching through the Google OAuth2 docs, but to summarize - Yes, there could be some discrepancies regarding the supported auth url parameters because we use `https://accounts.google.com/o/oauth2/auth` and not `https://accounts.google.com/o/oauth2/v2/auth` (_the token endpoint is also in a similar situation but the changes seems to be mostly for the OIDC compatibility and the end user shouldn't be affected in our case_).

So if `prompt=consent` doesn't work, try with `approval_prompt=force`.

I'll update the task in the roadmap for updating the Google `/v1/userinfo` endpoint with mentioning for the `/v2/auth` and `/v4/token` endpoints but for now it remains a low priority as I'll have to do more detailed search in the Google docs to find out what possible breaking changes it could introduce so that they could be eventually listed in the changelog.

In the meantime, if anyone manage to find more details related to the changes for the newer auth, token and userinfo endpoints, please list/link them.

**Reply by chnfyi** [reply]
I relied on the same answer you linked, on an [old google blog post](https://developers.googleblog.com/en/upcoming-changes-to-oauth-20-endpoint/), on the google documentation which explicitly assumes we use the `https://accounts.google.com/o/oauth2/v2/auth` base URI, and on previous experience.

I agree that it's confusing because they did many changes in the past years without making any announcements.

[Here](https://developers.google.com/identity/openid-connect/openid-connect) is the most complete page i found about the new implementation.

**Reply by blujedis** [reply]
For clarity, given I'm new to PB thought I'd mention this appears to be client side whereas the new suggested method `.authWithOAuth2Code()` is server side. Took me a min because newer examples don't show `.authWithOAuth2()` which is described above.

Lastly @ganigeorgiev is it a correct that in future versions we'll be able to use the .authWithOAuth2Code() server side or will we always need to take the client side route if we want the `refresh_token`. 

It appears this has been a topic a few times so I thought a quick post might clear this up :)

**Reply by ganigeorgiev** [reply]
@blujedis It is not clear what you are talking about.

Currently there are 2 ways to use OAuth2 in PocketBase - "All in one" flow (`authWithOAuth2`) and "Manual code-exchange" flow (`authWithOAuth2Code`), aka. the 2 tabs in https://pocketbase.io/docs/authentication/#authenticate-with-oauth2.

The "All in one" flow (`authWithOAuth2`) is the "newer" method and it was added for convenience to allow handling the OAuth2 flow in a single step without the need of page refresh or custom redirect handlers. Under the hood it calls `authWithOAuth2Code` but sends its response through the realtime connection that it was established with the initial call.

Please stop spamming and spend some time reading the documentation - https://pocketbase.io/docs/authentication/#authenticate-with-oauth2.

If it is still not clear, then please open a Q&A discussion and try to elaborate clearly what you are working on and what you don't understand but don't hijack unnecessary old discussions.
